src/tigerbeetle-java/src/main/java/com/tigerbeetle/Accounts.java

```diff
@@ -0,0 +1,317 @@
+     *
+     * @return an array of 16 bytes representing the 128-bit value.
+     * @throws IllegalStateException if not at a {@link #isValidPosition valid position}.
+     */
+    public byte[] getUserData() {
+        return getUInt128(at(Struct.UserData));
+    }
+
+    /**
+     * Gets an optional secondary identifier to link this account to an external entity.
+     *
+     * @param part a {@link UInt128} enum indicating which part of the 128-bit value is to be
+     *        retrieved.
+     * @return a {@code long} representing the the first 8 bytes of the 128-bit value if
+     *         {@link UInt128#LeastSignificant} is informed, or the last 8 bytes if
+     *         {@link UInt128#MostSignificant}.
+     * @throws IllegalStateException if not at a {@link #isValidPosition valid position}.
+     */
+    public long getUserData(final UInt128 part) {
+        return getUInt128(at(Struct.UserData), part);
+    }
+
+    /**
+     * Sets an optional secondary identifier to link this account to an external entity.
+     *
+     * @param leastSignificant a {@code long} representing the the first 8 bytes of the 128-bit
+     *        value.
+     * @param mostSignificant a {@code long} representing the the last 8 bytes of the 128-bit value.
+     * @throws IllegalStateException if not at a {@link #isValidPosition valid position}.
+     * @throws IllegalStateException if a {@link #isReadOnly() read-only} batch.
+     */
+    public void setUserData(final long leastSignificant, final long mostSignificant) {
+        putUInt128(at(Struct.UserData), leastSignificant, mostSignificant);
+    }
+
+    /**
+     * Sets an optional secondary identifier to link this account to an external entity.
+     * <p>
+     * May be zero, null values are converted to zero.
+     *
+     * @param userData an array of 16 bytes representing the 128-bit value.
+     * @throws IllegalArgumentException if {@code userData} is not 16 bytes long.
+     * @throws IllegalStateException if not at a {@link #isValidPosition valid position}.
+     * @throws IllegalStateException if a {@link #isReadOnly() read-only} batch.
+     */
+    public void setUserData(final byte[] userData) {
+        putUInt128(at(Struct.UserData), userData);
+    }
+
+    /**
+     * Gets an identifier that partitions the sets of accounts that can transact.
+     *
+     * @return a 32-bit integer.
+     * @throws IllegalStateException if not at a {@link #isValidPosition valid position}.
+     */
+    public int getLedger() {
+        return getUInt32(at(Struct.Ledger));
+    }
+
+    /**
+     * Sets an identifier that partitions the sets of accounts that can transact.
+     * <p>
+     * Must not be zero.
+     *
+     * @param ledger a 32-bit integer.
+     * @throws IllegalStateException if not at a {@link #isValidPosition valid position}.
+     * @throws IllegalStateException if a {@link #isReadOnly() read-only} batch.
+     */
+    public void setLedger(final int ledger) {
```

Comment:
With all of this API in general, why are we only internally storing as uints and not forcing the user to set/get our UInt types?

Reply:
To avoid object allocations, working with primitives is a very cheap operation. Java 20+ is coming with some nice improvements on primitive classes, that would help us with that.

In general, we don't want an object to be heap-allocated just for representing an unsigned 32-bit integer.

Reply:
Oh good point, thanks